universum/configuration_support.py

```diff
@@ -12,6 +15,203 @@
+
+..
+
+code_report
+    Basic usage is adding ``code_report=True`` to configuration description and ``--result-file="${CODE_REPORT_FILE}"``
+    to 'command' arguments.
+    Specifies step that performs static or syntax analysis of code.
+    Analyzers currently provided by Universum are: ``pylint``, ``svace`` and ``uncrustify``
+    (see `code_report parameters <code_report.html>`__ for details).
+
+.. _tc_tags:
+
+pass_tag, fail_tag
+    Basic usage is adding ``pass_tag="PASS", fail_tag="FAIL"`` to the configuration description.
+    These keys are currently implemented only for TeamCity build. You can specify one, both or neither of them
+    per step. Defining ``pass_tag="PASS"`` means that current build on TeamCity will be tagged with label ``PASS``
+    if this particular step succeeds. Defining ``fail_tag="FAIL"`` means that current build on TeamCity will be
+    tagged with label ``FAIL`` if this particular step fails. Key values can be set to any strings acceptable by
+    TeamCity as tags. It is not recommended to separate words in the tag with spaces, since you cannot create
+    a tag with spaces in TeamCity's web-interface. Every tag is added (if matching condition) after executing
+    build step it is set in, not in the end of all run.
+    ``pass_tag`` and ``fail_tag`` can also be used when
+    `multiplying build configuration <configuring.html#multiplying-build-configurations>`__, like this:
+
+    .. testsetup::
+
+        #!/usr/bin/env python3.7
+
+        from universum.configuration_support import Variations
+
+    .. testcode::
+
+        make = Variations([dict(name="Make ", command=["make"], pass_tag="pass_")])
+
+        target = Variations([
+            dict(name="Linux", command=["--platform", "Linux"], pass_tag="Linux"),
+            dict(name="Windows", command=["--platform", "Windows"], pass_tag="Windows")
+        ])
+
+        configs = make * target
+
+    This code will produce this list of configurations:
+
+    .. testcode::
+        :hide:
+
+        print("$ ./configs.py")
+        print(configs.dump())
+
+    .. testoutput::
+
+        $ ./configs.py
+        [{'name': 'Make Linux', 'command': 'make --platform Linux', 'pass_tag': 'pass_Linux'},
+        {'name': 'Make Windows', 'command': 'make --platform Windows', 'pass_tag': 'pass_Windows'}]
+
+    This means that tags "pass_Linux" and "pass_Windows" will be sent to TeamCity's build.
+
+
+.. note::
+
+    All the paths, specified in `command`, `artifacts` and `directory` parameters, can be
+    absolute or relative. All relative paths start from the project root (see :ref:`get_project_root`).
+
+    """
+    name: str
+    command: List[str]
+    environment: Dict[str, str]
+    artifacts: List[str]
+    report_artifacts: List[str]
+    artifact_prebuild_clean: bool
+    directory: str
+    critical: bool
+    background: bool
+    finish_background: bool
+    code_report: str
+    pass_tag: str
+    fail_tag: str
+    if_env_set: str
```

Comment:
1. Are you sure we are not limiting users to the listed key names?
Currently we are ignoring all unknown key names and, as fat as I know, some users already are using custom keys.
2. Can we avoid duplication between this list and ProjectConfigKeyType?